### PR TITLE
fix(map_height_fitter): return failure service response when ground fitting failed

### PR DESCRIPTION
**Pull request by KYabuuchi**
## Description
If ground fitting fails, return the service immediately as a failure.

Until now, when there was no point cloud map around the specified initial position, the given height (**basically z=0**) was returned as is.  If the correct height is not given, the NDT will fail to estimate. 

## Tests performed

Two tests were performed using logging_simulator and sample rosbag.

### How to test for case 1

```
# 1: Stop rosbag to prevent overwriting with correct gnss.
# 2: Set wrong gnss pose 
 ros2 topic pub /sensing/gnss/pose_with_covariance geometry_msgs/msg/PoseWithCovarianceStamped "{pose: {pose: {position: {x: 89435, y: 42047}}}}"
# 3: Call the initialization service
 ros2 service call /api/localization/initialize autoware_adapi_v1_msgs/srv/InitializeLocalization
```

After the above steps, a error message appears in the console indicating that ground fitting has failed.
Also, the localization status becomes `UNINITIALIZED`.

### How to test for case 2

```
# 1: Perform "2D Pose Estimate" in the area without a point cloud map as shown in the following figure.
```

After the above steps, a warning message appears in the console indicating that ground fitting has failed. 
The localization status does not change.

## Effects on system behavior

If ground fitting fails, the initial position estimation is aborted and a service failure is returned.
(Until now, initialization continued at the given elevation even if ground fitting failed.)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: map/map_height_fitter/src/map_height_fitter.cpp [hunk 1]
```diff
@@ -186,7 +192,14 @@ std::optional<Point> MapHeightFitter::Impl::fit(const Point & position, const st
     tf2::Transform transform{tf2::Quaternion{}, tf2::Vector3{}};
     tf2::fromMsg(stamped.transform, transform);
     point = transform * point;
-    point.setZ(get_ground_height(point));
+
+    //
```
Comment (isamu-takagi) [inline map/map_height_fitter/src/map_height_fitter.cpp:196]: Probably forgot to remove the comment.

File: map/map_height_fitter/src/map_height_fitter.cpp [hunk 2]
```diff
@@ -186,7 +192,14 @@ std::optional<Point> MapHeightFitter::Impl::fit(const Point & position, const st
...
+    auto point_optional = get_ground_height(point);
```
Comment (isamu-takagi) [inline map/map_height_fitter/src/map_height_fitter.cpp:197]: This value can be `const` and is not a `point`.

**Comment by isamu-takagi** [comment]
For case 2, no error when `enable_partial_load: false`. I think we have to change `get_ground_height`.
Screenshot from 2023-07-10 19-47-07

File: map/map_height_fitter/src/map_height_fitter.cpp (hunk 1 above)
Comment (KYabuuchi) [reply map/map_height_fitter/src/map_height_fitter.cpp]: Thank you! I removed it in 92be65d

File: map/map_height_fitter/src/map_height_fitter.cpp (hunk 2 above)
Comment (KYabuuchi) [reply map/map_height_fitter/src/map_height_fitter.cpp]: I made it const and modified the name in  [92be65d](https://github.com/autowarefoundation/autoware.universe/pull/4199/commits/92be65d6469a65c9de6546a861b84e4633e89845)

**Comment by KYabuuchi** [comment]
I have made a change so that `get_ground_height()` will fail if there are no points within a 50-meter (`max_search_radius`) range when determining the range of point cloud to consider.  Please see the line 154 to 157 the following link.

https://github.com/autowarefoundation/autoware.universe/blob/0171abbd1cdad18e0cc20dc47db0838a56ba783a/map/map_height_fitter/src/map_height_fitter.cpp#L140-L157

**Comment by KYabuuchi** [comment]
@isamu-takagi With the above changes, I believe the reported issue has been resolved.

> For case 2, no error when `enable_partial_load: false`. I think we have to change `get_ground_height`.

**Comment by KYabuuchi** [comment]
I found the related issue https://github.com/autowarefoundation/autoware.universe/issues/2048
@VRichardJP  If you have any concerns about this PR, please let us know. :+1:

**Review by isamu-takagi — APPROVED** [review]

**Comment by KYabuuchi** [comment]
I have noticed that there is a possibility that `YabLoc` and `Eagleye`, which are pose_estimator does not rely on point cloud maps, may not be able to complete initial pose estimation. :cry:

I will temporarily mark this PR as a draft. I will also consider an implementation that works with a general pose_estimator.

**Pull request closed without merging**